src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DepthFirstSearchHelper.java

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright (c) 2019, the Alpha Team.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1) Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2) Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package at.ac.tuwien.kr.alpha.common.depgraph;
+
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class DepthFirstSearchHelper {
+
+	private Set<Node> discoveredNodes;
+	private Deque<Node> finishedNodes;
+	private Map<Node, List<Node>> depthFirstForest;
+
+	private void reset() {
+		this.discoveredNodes = new HashSet<>();
+		this.finishedNodes = new LinkedList<>();
+		this.depthFirstForest = new HashMap<>();
+		this.depthFirstForest.put(null, new ArrayList<>());
+	}
+
+	public DfsResult performDfs(Map<Node, List<Edge>> nodes) {
+		return this.performDfs(nodes.keySet().iterator(), nodes);
+	}
+
+	/**
+	 * Performs a depth-first search on the given graph. The algorithm follows the approach outlined in "Introduction to
+	 * Algortihms, 3rd. Edition" by Cormen et al.
+	 * 
+	 * @param nodeVisitIt an Iterator defining in which sequence nodes should be visited
+	 * @param nodes       an adjacency map defining the dependency graph of an ASP program
+	 * @return a Set<Node> holding all finished nodes (i.e. all nodes at the end of the DFS run)
+	 */
+	public DfsResult performDfs(Iterator<Node> nodeVisitIt, Map<Node, List<Edge>> nodes) {
+		this.reset();
+		DfsResult retVal = new DfsResult();
+		Node tmp;
+		while (nodeVisitIt.hasNext()) {
```

Comment:
That wouldn't work, for-each loops require an Iterable to iterate over. An Iterator itself is not an Iterable (although I agree it would make sense...)